src/Microsoft.ML.Data/DataView/DetectAnomalyBySrCnnBatchTransform.cs

```diff
@@ -0,0 +1,332 @@
+                // The _lookAheadCursor's position should be on the last row of the previous batch (or -1).
+                Ch.Assert(_lastInBatchInLookAheadCursorDel());
+
+                var good = _lookAheadCursor.MoveNext();
+                // The two cursors should have the same number of elements, so if _input.MoveNext() returned true,
+                // then it must return true here too.
+                Ch.Assert(good);
+
+                do
+                {
+                    _inputGetterInLookAheadCursor(ref _currentInput);
+                    _parent.ProcessExample(_currentBatch, _currentInput);
+                } while (!_lastInBatchInLookAheadCursorDel() && _lookAheadCursor.MoveNext());
+
+                _parent.ProcessBatch(_currentBatch);
+                return true;
+            }
+        }
+    }
+
+    // TODO: SrCnn
+    public sealed class DetectAnomalyBySrCnnBatchTransform : BatchTransformBase<float, DetectAnomalyBySrCnnBatchTransform.Batch>
+    {
+        private readonly int _batchSize;
+
+        public DetectAnomalyBySrCnnBatchTransform(IHostEnvironment env, IDataView input, string inputColumnName, string outputColumnName, double threshold, int batchSize, double sensitivity, DetectMode detectMode)
+            : base(env, input, inputColumnName, outputColumnName, new VectorDataViewType(NumberDataViewType.Double, batchSize))
+        {
+            _batchSize = batchSize;
+        }
+
+        protected override Delegate[] CreateGetters(DataViewRowCursor input, Batch currentBatch, bool[] active)
+        {
+            if (!SchemaBindings.AnyNewColumnsActive(x => active[x]))
+                return new Delegate[1];
+            return new[] { currentBatch.CreateGetter(input, InputCol) };
+        }
+
+        protected override Batch InitializeBatch(DataViewRowCursor input) => new Batch(_batchSize);
+
+        protected override Func<bool> GetIsNewBatchDelegate(DataViewRowCursor input)
+        {
+            return () => input.Position % _batchSize == 0;
+        }
+
+        protected override Func<bool> GetLastInBatchDelegate(DataViewRowCursor input)
+        {
+            return () => (input.Position + 1) % _batchSize == 0;
+        }
+
+        protected override void ProcessExample(Batch currentBatch, float currentInput)
+        {
+            currentBatch.AddValue(currentInput);
+        }
+
+        protected override void ProcessBatch(Batch currentBatch)
+        {
+            currentBatch.Process();
+            currentBatch.Reset();
+        }
+
+        /// <summary>
+        /// The detect modes of SrCnn models.
+        /// </summary>
+        public enum DetectMode
+        {
+            /// <summary>
+            /// In this mode, output (IsAnomaly, RawScore, Mag).
+            /// </summary>
+            AnomalyOnly = 0,
+
+            /// <summary>
+            /// In this mode, output (IsAnomaly, AnomalyScore, Mag, ExpectedValue, BoundaryUnit, UpperBoundary, LowerBoundary).
```

Comment:
>AnomalyScore [](start = 49, length = 12)

Is this intentional that here it is called `AnomalyScore` and in the other modes it is called `RawScore`?

Reply:
Yes, in AnomalyAndMargin mode, we will have a better evaluated score based on sensitivity parameter, and it replaces RawScore as the final score.

---
In reply to: [424226362](https://github.com/mengaims/machinelearning/pull/1#discussion_r424226362) [](ancestors = 424226362)